### PR TITLE
25 Settings

### LINKED ISSUE #25: Implement Settings class
The settings class will contain all settings that can be set by the user for the extension.

### PR DESCRIPTION
**Pull request by mdingjan**
Will close #25 

Two 'main concepts' have been added within this PR: a frontend for the Settings and a backend.

Frontend:
- Options.html has been added; it contains for the moment the most important possible preferences of the extension.
- The Settings page has been styled by @thervh70. He did a great job there. :)
- Options.js has been added; it contains the functions that are needed for interaction on the Settings page (e.g. storing and restoring saved preferences of the user).

Backend:
- A Settings class has been added which basically retrieves that settings that have been stored within the chrome storage.
- An InternalSettings class has been added, which is yet empty. This class will contain backend settings (thus internal settings) for the extension and will be an issue for a future sprint.

So how can you review this PR?
- Please check the code.
- Please interact with the settings page of Octopeer, and check if you can break it and notify me whether you like it. :)

### CODE REVIEW

File: resources/options.html [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                    <li>
+                        <input type="checkbox" id="tabs" checked>
+                        <label for="tabs">other active tabs that are used.</label>
+                    </li><li>
+                        <input type="checkbox" id="comments" checked>
+                        <label for="comments"> the comments that I add to a pull request. </label>
+                    </li><li>
+                        <input type="checkbox" id="peer_comments" checked>
+                        <label for="peer_comments"> the comments of my peers that have been added to a pull request. </label>
+                    </li><li>
```
Comment (mpsijm) [inline resources/options.html]: We do not have a style check for HTML, but can I be picky and want the `</li><li>` split with a line break? :P

File: src/octopeer-github/main/Settings/Settings.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+     * Gets the logging preference, from the chrome storage.
+     * @returns {Boolean}
+     */
+    public getLogging() {
+        chrome.storage.sync.get("loggingEnabled", function (obj) {
+            this.logging = obj;
+        });
+        return this.logging;
```
Comment (mpsijm) [inline src/octopeer-github/main/Settings/Settings.ts]: Does this work? As a callback usually indicates that the call is asynchronous and thus is not executed immediately, meaning `this.logging` is only set after this method has returned

**Comment by mpsijm** [comment]
As a user
I want to click on everything as much as I can
So I can break everything :D

JK, it's not that bad, but maybe you want to think about reducing the amount of popups :)

Gyazo GIF: settings page

**Comment by mpsijm** [comment]
Even a better idea, why have a "submit" button at all? You can also save on every change, that also solves the problem above :)

File: resources/options.html (hunk 1 above)
Comment (mdingjan) [inline resources/options.html]: Sure xD Will fix :)

File: resources/options.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
+/**
+ * Created by Mitchell on 10-5-2016.
+ * Contains necessary functions for the Settings page of Octopeer.
+ */
+
+// Helper function; simplifies the calls to specific document e-L-ements.
+function l(id) {
+    return document.getElementById(id)
```
Comment (MathiasMeuleman) [inline resources/options.js]: Why use plain JavaScript when we have jQuery in our project?

Comment (mpsijm) [inline resources/options.js]: I used this so much I didn't even notice XD But indeed, jQuery is even already loaded into the page :)

File: src/octopeer-github/main/Settings/Settings.ts (hunk 2 above)
Comment (mdingjan) [inline src/octopeer-github/main/Settings/Settings.ts]: I get your point. I actualy need a clean way of testing this..

**Comment by mdingjan** [comment]
@mpsijm About the submit button, we can indeed do this without the popups and show inline that changes have been saved. 
It might be a point of intrest to discuss with the team, in relation to styling it.
Btw, nice user story xD

File: resources/options.html [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                    <li>
+                        <input type="checkbox" id="tabs" checked>
+                        <label for="tabs">other active tabs that are used.</label>
+                    </li> <li>
+                        <input type="checkbox" id="comments" checked>
+                        <label for="comments"> the comments that I add to a pull request. </label>
+                    </li> <li>
+                        <input type="checkbox" id="peer_comments" checked>
+                        <label for="peer_comments"> the comments of my peers that have been added to a pull request. </label>
+                    </li> <li>
```
Comment (mpsijm) [inline resources/options.html]: I'm very sorry to say that this is not a line break... :P

Comment (mdingjan) [inline resources/options.html]: My bad.., thanks for reviewing (A)

File: src/octopeer-github/main/Settings/Settings.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
+/**
+ * Created by Mitchell on 12-5-2016.
+ */
+
+/**
+ * Class from which user settings can be retrieved.
+ * These settings have been maintained by the front-end.
+ */
+class Settings {
```
Comment (MathiasMeuleman) [inline src/octopeer-github/main/Settings/Settings.ts]: Wasn't this supposed to be a Singleton class? (If I remember correctly from the UML). I would expect a getInstance() function

Comment (mpsijm) [inline src/octopeer-github/main/Settings/Settings.ts]: Nooonononono :P I bet typescript has much better ways of creating singletons, I haven't seen much of it, but I'd first research that before using `getInstance()` :)

Comment (mpsijm) [inline src/octopeer-github/main/Settings/Settings.ts]: Like this: [Namespace equivalent of `getInstance()`](http://stackoverflow.com/a/30174360/3227948)

Comment (mdingjan) [inline src/octopeer-github/main/Settings/Settings.ts]: Done, see commit 73ce230 below

File: resources/options.js (hunk 3 above)
Comment (mdingjan) [inline resources/options.js]: Done, see commit 73ce230 below

File: src/octopeer-github/main/Settings/Settings.ts (hunk 5 above)
Comment (MathiasMeuleman) [inline src/octopeer-github/main/Settings/Settings.ts]: Oké, yes. Could you please have a look at that @mdingjan? The changes are well done! This is the only thing in the PR bothering me

Comment (mpsijm) [inline src/octopeer-github/main/Settings/Settings.ts]: I don't see this in that commit? I see you've resolved other feedback though, so that's good :)

Comment (MathiasMeuleman) [inline src/octopeer-github/main/Settings/Settings.ts]: Still to come @mpsijm. @mdingjan is quick, but not 20-seconds quick ;)

Comment (mdingjan) [inline src/octopeer-github/main/Settings/Settings.ts]: Done, you can find our first Singleton class in commit 79cf6a3 :)

**Comment by mpsijm** [comment]
Looks good to me, if Mathias is okay with this, he can merge :)

**Comment by MathiasMeuleman** [comment]
Much better! Tnx, will merge

**Pull request merged**